### PR TITLE
Adding SSL support to server and command line clients.

### PR DESCRIPTION
**Pull request by dkrupp**

### CODE REVIEW

File: libcodechecker/libclient/authentication_helper.py [hunk 1]
```diff
@@ -20,10 +20,12 @@
...
-    def __init__(self, host, port, uri, session_token=None):
+    def __init__(self, protocol, host, port, uri,
+                 session_token=None):
         self.__host = host
         self.__port = port
-        self.transport = THttpClient.THttpClient(self.__host, self.__port, uri)
+        url = protocol+"://"+host+":"+str(port)+uri
+        self.transport = THttpClient.THttpClient(url)
         self.protocol = TJSONProtocol.TJSONProtocol(self.transport)
         self.client = codeCheckerAuthentication.Client(self.protocol)
```
Comment (Xazax-hun) [inline libcodechecker/libclient/authentication_helper.py]: The `protocol` and `self.protocol` are two distinct things. Maybe we should name them separately.

File: libcodechecker/libclient/product_helper.py [hunk 2]
```diff
@@ -19,10 +19,11 @@
 
 
 class ThriftProductHelper(object):
-    def __init__(self, host, port, uri, session_token=None):
+    def __init__(self, protocol, host, port, uri, session_token=None):
         self.__host = host
         self.__port = port
-        self.transport = THttpClient.THttpClient(self.__host, self.__port, uri)
+        url = protocol+"://"+host+":"+str(port)+uri
```
Comment (Xazax-hun) [inline libcodechecker/libclient/product_helper.py]: This url creating code is there multiple times, maybe worth to create a function?

Comment (Xazax-hun) [reply libcodechecker/libclient/product_helper.py]: Same comment about the name of `protocol` as above.

File: libcodechecker/server/client_db_access_server.py [hunk 3]
```diff
@@ -585,6 +586,22 @@ def __init__(self,
...
+            if os.path.isfile(ssl_key_file) and os.path.isfile(ssl_cert_file):
+                LOG.info("Initiating SSL. Server listening on secure socket.")
+                LOG.debug("Using cert file:"+ssl_cert_file)
+                LOG.debug("Using key file:"+ssl_key_file)
+                self.socket = ssl.wrap_socket(self.socket, server_side=True,
+                                              keyfile=ssl_key_file,
+                                              certfile=ssl_cert_file)
+
+            else:
+                LOG.info("SSL key or cert files not found. cert:" +
```
Comment (Xazax-hun) [inline libcodechecker/server/client_db_access_server.py]: Do we want to display this message even if the user never configured this? It is just an ergonomic change in the future.

File: libcodechecker/util.py [hunk 4]
```diff
@@ -266,7 +266,7 @@ def split_server_url(url):
     LOG.debug("Parsing server url '{0}'".format(url))
 
     protocol = 'http'
-    if url.startswith('http'):
+    if url.startswith('http') or url.startswith('https'):
```
Comment (Xazax-hun) [inline libcodechecker/util.py]: If url starts with `https` it also starts with `http` right?  This check might be redundant.

File: tests/libtest/thrift_client_to_db.py [hunk 5]
```diff
@@ -95,16 +95,18 @@ def __exit__(self, type, value, tb):
...
         from libcodechecker import session_manager
         from codeCheckerDBAccess_v6 import codeCheckerDBAccess
         from codeCheckerDBAccess_v6.constants import MAX_QUERY_SIZE
 
         self.max_query_size = MAX_QUERY_SIZE
-        transport = THttpClient.THttpClient(
-            host, port, '/' + product + '/v' + VERSION + endpoint)
+        url = protocol+"://"+host+":"+str(port) +\
+            '/' + product + '/v' + VERSION + endpoint
+        print ("setup viewer client:"+url)
```
Comment (Xazax-hun) [inline tests/libtest/thrift_client_to_db.py]: Is this print intended? If so, maybe having a space after colon and starting with capital letter would be great.

File: tests/libtest/codechecker.py [hunk 6]
```diff
@@ -405,19 +408,20 @@ def remove_test_package_product(test_folder, check_env=None):
         server_data['check_env'] = check_env
 
     # Clean the previous session if any exists.
-    logout(server_data, test_folder)
+    logout(server_data, test_folder, protocol)
 
+    url = protocol + "://"+server_data['viewer_host'] +\
```
Comment (Xazax-hun) [inline tests/libtest/codechecker.py]: Maybe you could use the just introduced util here?

File: tests/libtest/thrift_client_to_db.py [hunk 7]
```diff
@@ -95,16 +95,18 @@ def __exit__(self, type, value, tb):
...
                  auto_handle_connection=True, session_token=None):
         # Import only if necessary; some tests may not add this to PYTHONPATH.
         from libcodechecker import session_manager
         from codeCheckerDBAccess_v6 import codeCheckerDBAccess
         from codeCheckerDBAccess_v6.constants import MAX_QUERY_SIZE
 
         self.max_query_size = MAX_QUERY_SIZE
-        transport = THttpClient.THttpClient(
-            host, port, '/' + product + '/v' + VERSION + endpoint)
+        url = protocol+"://"+host+":"+str(port) +\
```
Comment (Xazax-hun) [inline tests/libtest/thrift_client_to_db.py]: Is the introduced util applicable here?

File: tests/libtest/thrift_client_to_db.py [hunk 8]
```diff
@@ -171,17 +173,17 @@ def __getattr__(self, attr):
...
         # Import only if necessary; some tests may not add this to PYTHONPATH.
         from libcodechecker import session_manager
         from ProductManagement_v6 import codeCheckerProductService
-
         full_uri = '/v' + VERSION + uri
         if product:
             full_uri = '/' + product + full_uri
-
-        transport = THttpClient.THttpClient(host, port, full_uri)
+        url = proto+"://"+host+":"+str(port)+full_uri
```
Comment (Xazax-hun) [inline tests/libtest/thrift_client_to_db.py]: Is the introduced util applicable here?

File: tests/libtest/thrift_client_to_db.py [hunk 9]
```diff
@@ -149,13 +151,13 @@ def _getAll_emu(self, func_name, *args):
...
 
-    def __init__(self, host, port, uri, auto_handle_connection=True,
+    def __init__(self, proto, host, port, uri, auto_handle_connection=True,
                  session_token=None):
         # Import only if necessary; some tests may not add this to PYTHONPATH.
         from libcodechecker import session_manager
         from Authentication_v6 import codeCheckerAuthentication
-
-        transport = THttpClient.THttpClient(host, port, '/v' + VERSION + uri)
+        url = proto+"://"+host+":"+str(port)+'/v' + VERSION + uri
```
Comment (Xazax-hun) [inline tests/libtest/thrift_client_to_db.py]: Is the introduced util applicable here?

File: tests/libtest/codechecker.py (hunk 6 above)
Comment (whisperity) [reply tests/libtest/codechecker.py]: Please format the code better so `+` is wrapped in spaces around the operand.

File: docs/products.md [hunk 10]
```diff
@@ -17,10 +17,11 @@ database is an isolated storage of results, completely separate, and can be
...
 
 Command-line tools that are used to access analysis data take an `--url
-PRODUCT_URL` parameter, which specifies the server host, port, and the
-product's unique endpoint in the following format: `localhost:8001/Default`,
-where `Default` is the product's endpoint. See the User guide for further
-details.
+PRODUCT_URL` parameter, which specifies the access protocol, 
+server host, port, and the product's unique endpoint 
+in the following format: `http[s]://localhost:8001/Default`,
+where `Default` is the product's endpoint. See the [User Guide](/docs/user_guide.md#product_url-format )
```
Comment (whisperity) [inline docs/products.md]: There is a space before the `)`

File: docs/usage.md [hunk 11]
```diff
@@ -69,10 +69,19 @@ create a workspace directory, where the database will be stored.
  mkdir ./ws
  CodeChecker server -w ./ws -v 8555
 ```
-2. Store the results in the server under run name "tmux":
+A default product called `Default` will be automatically created where you can store your results.
+
+2. Store the results in the server under run name "tmux" (in the `Default` product):
 ```
- CodeChecker store ./reports --host localhost --port 8555 --name tmux
+ CodeChecker store ./reports --name tmux --url http://loclahost:8555/Default
```
Comment (whisperity) [inline docs/usage.md]: `localhost` has a typo in it

File: docs/products.md [hunk 12]
```diff
@@ -17,10 +17,11 @@ database is an isolated storage of results, completely separate, and can be
 added and removed from a server on the fly.
 
 Command-line tools that are used to access analysis data take an `--url
-PRODUCT_URL` parameter, which specifies the server host, port, and the
-product's unique endpoint in the following format: `localhost:8001/Default`,
-where `Default` is the product's endpoint. See the User guide for further
-details.
+PRODUCT_URL` parameter, which specifies the access protocol, 
+server host, port, and the product's unique endpoint 
+in the following format: `http[s]://localhost:8001/Default`,
```
Comment (whisperity) [inline docs/products.md]: Is `http` **mandatory** even if we are using only HTTP?

File: docs/user_guide.md [hunk 13]
```diff
@@ -940,6 +967,15 @@ root account arguments:
...
 ~~~~~~~~~~~~~~~~~~~~~
 
+### Enfore secure socket (SSL)
+
+You can enforce SSL security on your listening socket. In this case all clients must
+access your server using the `https://host:port` URL format.
+
+To enable SSL simply place ssl certificate to `<CONFIG_DIRECTORY>/cert.pem`
+and the corresponding private key to `<CONFIG_DIRECTORY>/key.pem`.
+When the server finds these files upon start-up. SSL will be automatically enabled.
```
Comment (whisperity) [inline docs/user_guide.md]: Sentences are broken wrongly, perhaps you meant a `,` instead of `.`?

File: docs/user_guide.md [hunk 14]
```diff
@@ -940,6 +967,15 @@ root account arguments:
                         not require authentication otherwise.
 ~~~~~~~~~~~~~~~~~~~~~
 
+### Enfore secure socket (SSL)
+
+You can enforce SSL security on your listening socket. In this case all clients must
+access your server using the `https://host:port` URL format.
+
+To enable SSL simply place ssl certificate to `<CONFIG_DIRECTORY>/cert.pem`
```
Comment (whisperity) [inline docs/user_guide.md]: SSL is uppercase, and then lowercase.

File: tests/functional/ssl/__init__.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    TEST_WORKSPACE = env.get_workspace('ssl')
+
+    os.environ['TEST_WORKSPACE'] = TEST_WORKSPACE
+
+    test_config = {}
+
+    # Setup environment variables for the test cases.
+    host_port_cfg = {'viewer_host': 'localhost',
+                     'viewer_port': env.get_free_port(),
+                     'viewer_product': 'authentication'}
```
Comment (whisperity) [inline tests/functional/ssl/__init__.py]: Name of the test package is `ssl` and not `authentication`!!!

File: tests/functional/ssl/__init__.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+    codechecker_cfg = {
+        'check_env': test_env,
+        'workspace': TEST_WORKSPACE,
+        'checkers': []
+    }
+
+    codechecker_cfg.update(host_port_cfg)
+
+    codechecker_cfg['run_names'] = []
```
Comment (whisperity) [inline tests/functional/ssl/__init__.py]: Are we using this variable for anything later on?

File: tests/functional/ssl/__init__.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+def teardown_package():
+    """Stop the CodeChecker server and clean up after the tests."""
+    # TODO If environment variable is set keep the workspace
+    # and print out the path.
+    global TEST_WORKSPACE
+
+    # Removing the product through this server requires credentials.
+    codechecker_cfg = env.import_test_cfg(TEST_WORKSPACE)['codechecker_cfg']
+    codechecker.remove_test_package_product(TEST_WORKSPACE,
```
Comment (whisperity) [inline tests/functional/ssl/__init__.py]: If you didn't call `add_test_package_product()` in `startUp()`, you should not (need not) call `remove_test_package_product` here.

File: tests/libtest/codechecker.py [hunk 18]
```diff
@@ -34,15 +34,16 @@ def wait_for_postgres_shutdown(workspace):
...
+          protocol='http'):
     """
     Perform a command-line login to the server.
     """
     print("Logging in")
     port = str(codechecker_cfg['viewer_port'])
     login_cmd = ['CodeChecker', 'cmd', 'login', username,
                  '--verbose', 'debug',
-                 '--url', 'localhost:' + port]
+                 '--url', protocol+'://' + 'localhost:' + port]
```
Comment (whisperity) [inline tests/libtest/codechecker.py]: Format operands better (spaces around operand)

File: tests/libtest/codechecker.py [hunk 19]
```diff
@@ -87,7 +88,7 @@ def logout(codechecker_cfg, test_project_path):
     logout_cmd = ['CodeChecker', 'cmd', 'login',
                   '--logout',
                   '--verbose', 'debug',
-                  '--url', 'localhost:' + port]
+                  '--url', protocol+'://'+'localhost:' + port]
```
Comment (whisperity) [inline tests/libtest/codechecker.py]: operator format

File: tests/libtest/codechecker.py [hunk 20]
```diff
@@ -351,12 +353,13 @@ def add_test_package_product(server_data, test_folder, check_env=None):
     codechecker_cfg.update(server_data)
 
     # Clean the previous session if any exists.
-    logout(codechecker_cfg, test_folder)
+    logout(codechecker_cfg, test_folder, protocol)
 
+    url = protocol + "://"+server_data['viewer_host'] +\
```
Comment (whisperity) [inline tests/libtest/codechecker.py]: operator format

File: tests/ssl_example_cert/cert.pem [hunk 21]
```diff
@@ -0,0 +1,32 @@
+-----BEGIN CERTIFICATE-----
```
Comment (whisperity) [inline tests/ssl_example_cert/cert.pem]: Is this a self-signed certificate? Are we sure we are not leaking anything here?

File: docs/user_guide.md [hunk 22]
```diff
@@ -940,6 +967,15 @@ root account arguments:
                         not require authentication otherwise.
 ~~~~~~~~~~~~~~~~~~~~~
 
+### Enfore secure socket (SSL)
```
Comment (whisperity) [inline docs/user_guide.md]: Perhaps we should link to some generally available guide which explains how to make an SSL certificate?

File: tests/ssl_example_cert/cert.pem (hunk 21 above)
Comment (whisperity) [inline tests/ssl_example_cert/cert.pem]: The shell actions that were used to make these test files, I think, should be saved to this folder in an `.sh` file too.

**Pull request merged**